File: app/src/main/java/com/exemple/habitapp/NotificationHelper.java
package com.exemple.habitapp;

import android.Manifest;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Build;

import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import androidx.core.content.ContextCompat;

public final class NotificationHelper {

    public static final String CHANNEL_REMINDERS = "habit_reminders";

    private NotificationHelper() {
    }

    public static void createChannels(Context context) {
        if (Build.VERSION.SDK_INT < 26) return;

        NotificationChannel channel = new NotificationChannel(
                CHANNEL_REMINDERS,
                "Lembretes do HabitApp",
                NotificationManager.IMPORTANCE_DEFAULT
        );
        channel.setDescription("Lembretes inteligentes de agua, foco e rotina.");

        NotificationManager manager = context.getSystemService(NotificationManager.class);
        if (manager != null) {
            manager.createNotificationChannel(channel);
        }
    }

    public static void showReminder(Context context, String type) {
        if (Build.VERSION.SDK_INT >= 33
                && ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
            return;
        }

        createChannels(context);
        SharedPreferences prefs = context.getSharedPreferences("habit_data", Context.MODE_PRIVATE);
        HabitStore.ensureToday(prefs);

        String title = getTitle(type);
        String message = getMessage(type, prefs);
        PendingIntent pendingIntent = getPendingIntent(context, type);

        NotificationCompat.Builder builder = new NotificationCompat.Builder(context, CHANNEL_REMINDERS)
                .setSmallIcon(R.drawable.ic_notification)
                .setContentTitle(title)
                .setContentText(message)
                .setStyle(new NotificationCompat.BigTextStyle().bigText(message))
                .setContentIntent(pendingIntent)
                .setAutoCancel(true)
                .setPriority(NotificationCompat.PRIORITY_DEFAULT)
                .setColor(ContextCompat.getColor(context, getColor(type)))
                .addAction(R.drawable.ic_notification, getActionLabel(type), pendingIntent);

        NotificationManagerCompat.from(context).notify(type.hashCode(), builder.build());
    }

    private static PendingIntent getPendingIntent(Context context, String type) {
        Intent intent = new Intent(context, MainActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
        intent.putExtra("nav_target", getDestination(type));
        return PendingIntent.getActivity(
                context,
                type.hashCode(),
                intent,
                PendingIntent.FLAG_UPDATE_CURRENT | immutableFlag()
        );
    }

    private static String getTitle(String type) {
        if (ReminderScheduler.TYPE_WATER.equals(type)) return "Hora da agua";
        if (ReminderScheduler.TYPE_FOCUS.equals(type)) return "Bloco de foco";
        return "Fechamento da rotina";
    }

    private static String getMessage(String type, SharedPreferences prefs) {
        if (ReminderScheduler.TYPE_WATER.equals(type)) {
            int falta = Math.max(0, HabitStore.getMetaAguaMl(prefs) - HabitStore.getAguaMl(prefs));
            return falta == 0
                    ? "Meta de agua fechada. Abra o painel para manter a sequencia."
                    : "Faltam " + falta + " ml para bater a meta de hoje. Um copo agora ja ajuda.";
        }

        if (ReminderScheduler.TYPE_FOCUS.equals(type)) {
            int falta = Math.max(0, prefs.getInt("meta_estudos_min", 60) - prefs.getInt("estudos_concluidos_min", 0));
            return falta == 0
                    ? "Foco do dia fechado. Revise seu progresso e proteja a rotina."
                    : "Faltam " + falta + " min de foco hoje. Comece com uma sessao curta.";
        }

        int checklist = HabitStore.getChecklistConcluido(prefs, HabitStore.todayKey());
        return checklist >= 3
                ? "Checklist completo. Veja suas conquistas e prepare o proximo dia."
                : "Faltam " + (3 - checklist) + " itens do checklist para fechar a rotina.";
    }

    private static int getDestination(String type) {
        if (ReminderScheduler.TYPE_WATER.equals(type)) return R.id.agua;
        if (ReminderScheduler.TYPE_FOCUS.equals(type)) return R.id.estudos;
        return R.id.rotina;
    }

    private static int getColor(String type) {
        if (ReminderScheduler.TYPE_WATER.equals(type)) return R.color.water;
        if (ReminderScheduler.TYPE_FOCUS.equals(type)) return R.color.study;
        return R.color.success;
    }

    private static String getActionLabel(String type) {
        if (ReminderScheduler.TYPE_WATER.equals(type)) return "Registrar agua";
        if (ReminderScheduler.TYPE_FOCUS.equals(type)) return "Abrir foco";
        return "Ver rotina";
    }

    public static int immutableFlag() {
        return Build.VERSION.SDK_INT >= 23 ? PendingIntent.FLAG_IMMUTABLE : 0;
    }
}

File: app/src/main/java/com/exemple/habitapp/notifications/NotificationHelper.kt
package com.exemple.habitapp.notifications

import android.Manifest
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import com.exemple.habitapp.MainActivity
import com.exemple.habitapp.R
import com.exemple.habitapp.data.HabitRepository
import com.exemple.habitapp.ui.navigation.HabitRoute
import com.exemple.habitapp.widgets.WidgetActionReceiver

object NotificationHelper {
    private const val CHANNEL_ID = "habit_reminders"

    fun createChannels(context: Context) {
        if (Build.VERSION.SDK_INT < 26) return
        val manager = context.getSystemService(NotificationManager::class.java)
        val channel = NotificationChannel(
            CHANNEL_ID,
            "Lembretes do HabitApp",
            NotificationManager.IMPORTANCE_DEFAULT,
        ).apply {
            description = "Lembretes inteligentes de água, foco e rotina."
        }
        manager.createNotificationChannel(channel)
    }

    fun showReminder(context: Context, type: String) {
        if (Build.VERSION.SDK_INT >= 33 &&
            ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED
        ) {
            return
        }
        createChannels(context)
        val state = HabitRepository(context).snapshot()
        val (title, message) = when (type) {
            ReminderScheduler.TYPE_WATER -> {
                val missing = (state.waterGoalMl - state.waterMl).coerceAtLeast(0)
                "Hora da água" to if (missing == 0) {
                    "Meta de água fechada. Abra o painel para manter a sequência."
                } else {
                    "Faltam $missing ml para bater a meta de hoje. Um copo agora já ajuda."
                }
            }
            ReminderScheduler.TYPE_FOCUS -> {
                val missing = (state.focusGoalMinutes - state.focusMinutes).coerceAtLeast(0)
                "Bloco de foco" to if (missing == 0) {
                    "Foco do dia fechado. Revise seu progresso e proteja a rotina."
                } else {
                    "Faltam $missing min de foco hoje. Comece com uma sessão curta."
                }
            }
            else -> {
                val missing = 4 - state.routineBlocks.count { it }
                "Fechamento da rotina" to if (missing <= 0) {
                    "Rotina completa. Veja conquistas e prepare o próximo dia."
                } else {
                    "Faltam $missing blocos para fechar a rotina de hoje."
                }
            }
        }

        val pendingIntent = openAppIntent(context, type)
        val notification = NotificationCompat.Builder(context, CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_notification)
            .setContentTitle(title)
            .setContentText(message)
            .setStyle(NotificationCompat.BigTextStyle().bigText(message))
            .setAutoCancel(true)
            .setContentIntent(pendingIntent)
            .addAction(R.drawable.ic_notification, actionLabel(type), actionIntent(context, type))
            .addAction(R.drawable.ic_clock_history, "Lembrar em 15 min", snoozeIntent(context, type))
            .addAction(R.drawable.ic_nav_home, "Abrir app", pendingIntent)
            .build()
        NotificationManagerCompat.from(context).notify(type.hashCode(), notification)
    }

    private fun openAppIntent(context: Context, type: String): PendingIntent {
        val intent = Intent(context, MainActivity::class.java)
            .putExtra(MainActivity.EXTRA_NAV_ROUTE, destinationRoute(type))
            .setFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP)
        return PendingIntent.getActivity(
            context,
            type.hashCode(),
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
    }

    private fun actionIntent(context: Context, type: String): PendingIntent {
        val action = when (type) {
            ReminderScheduler.TYPE_WATER -> WidgetActionReceiver.ACTION_ADD_WATER
            ReminderScheduler.TYPE_FOCUS -> WidgetActionReceiver.ACTION_ADD_FOCUS
            else -> WidgetActionReceiver.ACTION_COMPLETE_ROUTINE
        }
        val intent = Intent(context, WidgetActionReceiver::class.java).setAction(action)
        return PendingIntent.getBroadcast(
            context,
            410 + type.hashCode(),
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
    }

    private fun snoozeIntent(context: Context, type: String): PendingIntent {
        val intent = Intent(context, WidgetActionReceiver::class.java)
            .setAction(WidgetActionReceiver.ACTION_SNOOZE_REMINDER)
            .putExtra(ReminderScheduler.EXTRA_TYPE, type)
        return PendingIntent.getBroadcast(
            context,
            520 + type.hashCode(),
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
    }

    private fun destinationRoute(type: String): String = when (type) {
        ReminderScheduler.TYPE_WATER -> HabitRoute.Water.route
        ReminderScheduler.TYPE_FOCUS -> HabitRoute.Focus.route
        else -> HabitRoute.Routine.route
    }

    private fun actionLabel(type: String): String = when (type) {
        ReminderScheduler.TYPE_WATER -> "Registrar água"
        ReminderScheduler.TYPE_FOCUS -> "Registrar foco"
        else -> "Marcar rotina"
    }
}
